macros/contexts/contextUnits.pl

```diff
@@ -0,0 +1,2056 @@
+			sameUnits          => $options{sameUnits}          // 0,
+			partialCredit      => $options{partialCredit}      // .5,
+			factorUnits        => $options{factorUnits}        // 1,
+		},
+		context => 'Context'
+	);
+}
+
+#
+#  Create the Units and LimitedUnits contexts, and the
+#  Unit(), NumberWithUnit(), and FormulaWithUnit() functions.
+#
+sub Init {
+	$main::context{Units}        = context::Units::extending("Numeric");
+	$main::context{LimitedUnits} = context::Units::extending("LimitedNumeric", limited => 1);
+	sub main::Unit            { Value->Package("Unit()")->new(@_) }
+	sub main::NumberWithUnits { Value->Package("NumberWithUnit()")->new(@_) }
+
+	sub main::FormulaWithUnits {
+		return Value->Package("Formula()")->new(Value->Package("NumberWithUnit()")->new(@_));
+	}
+}
+
+#################################################################################################
+#################################################################################################
+
+#
+#  The context subclass that adds unit-handling functions
+#
+package context::Units::Context;
+our @ISA = ('Parser::Context');
+
+#
+#  The units from the original Units package
+#
+our %UNITS = (%Units::known_units);
+$UNITS{$_} = $UNITS{L} for ('litre', 'litres', 'litre', 'litres');    # add these extras
+
+#
+#  The categories of units that can be selected.
+#
+#  These give the fundamental units of the unit names to be added to
+#  the context, or a list of such, or a list of names of known units.
+#  If a name begins with a dash, then REMOVE the category or named
+#  unit.  For example, the "length" category excludes the lengths that
+#  are part of the "atomics" and "astronomy" categories.  If a name
+#  ends in an asterisk, then add or remove all the aliases for that
+#  unit as well.
+#
+our %categories = (
+	angles            => { rad => 1 },
+	time              => { s   => 1 },
+	length            => [ { m => 1 }, '-atomics', '-astronomy' ],
+	"metric-length"   => [ { m => 1 }, '-atomics', '-astronomy', '-imperial-length' ],
+	"imperial-length" => [ 'in*', 'ft*', 'mi*', 'furlong*' ],
+	volume            => { m => 3 },
+	velocity          => { m => 1, s => -1 },
+	mass              => [ { kg => 1 }, '-astronomy' ],
+	temperature       => [ { degC => 1 }, { defF => 1 }, { K => 1 } ],
+	frequency         => { rad => 1, s => -1 },
+	force             => { m => 1, kg => 1, s => -2 },
+	energy            => { m => 2, kg => 1, s => -2 },
+	power             => [ { m => 2, kg => 1, s => -3 }, '-astronomy' ],
+	pressure          => { m => -1, kg => 1, s => -2 },
+	electricity       => [
+		{ amp => 1 },
+		{ amp => 1,  s => 1 },
+		{ kg  => 1,  m => 2,  amp => -1, s => -3 },
+		{ kg  => -1, m => -2, amp => 1,  s => 3 },
+		{ amp => 2,  s => 4,  kg  => -1, m => -2 },
+		{ kg  => 1,  m => 2,  amp => -2, s => -3 },
+		{ kg  => -1, m => -2, amp => 2,  s => 3 },
+	],
+	magnatism   => [ { kg => 1, amp => -1, s => -2 }, { kg => 1, m => 2, amp => -1, s => -2 }, ],
```

Comment:
Shouldn't this be `magnetism`?

Reply:
OOPS!  Thanks for catching that!  I will fix.